client/app/visualizations/chart/ChartEditor.jsx

```diff
@@ -0,0 +1,668 @@
+                      onChange={this.toggleReverseY}
+                      checked={opts.reverseY || false}
+                    />
+                    <i className="input-helper" /> Reverse Order
+                  </label>
+                </div>
+
+              </div>
+            </Tabs.TabPane> : null }
+          {opts.globalSeriesType !== 'custom' ?
+            <Tabs.TabPane key="series" tab="Series">
+              <ChartSeriesEditor
+                seriesOptions={opts.seriesOptions}
+                seriesList={seriesList}
+                type={opts.globalSeriesType}
+                updateSeriesList={this.updateSeriesList}
+                updateSeriesOptions={this.updateSeriesOptions}
+                clientConfig={this.props.clientConfig}
+              />
+            </Tabs.TabPane> : null }
+          {opts.globalSeriesType === 'heatmap' ?
+            <Tabs.TabPane key="colors" tab="Colors">
+              <div className="form-group">
+                <label className="control-label">Color Scheme</label>
+                <Select
+                  value={opts.colorScheme}
+                  placeholder="Choose color scheme..."
+                  onChange={this.updateColorScheme}
+                  allowClear
+                  dropdownClassName="ant-dropdown-in-bootstrap-modal"
+                >
+                  {map(colorSchemes, c => <Select.Option key={c}>{c}</Select.Option>)}
+                </Select>
+              </div>
+              <div className="row">
+                <div className="col-xs-6">
+                  {opts.colorScheme === 'Custom...' ?
+                    <div className="form-group">
+                      <label className="control-label">Min Color</label>
+                      <ColorSelect
+                        value={opts.heatMinColor}
+                        onChange={this.updateHeatMinColor}
+                      />
+                    </div> : null }
+                </div>
+                <div className="col-xs-6">
+                  {opts.colorScheme === 'Custom...' ?
+                    <div className="form-group">
+                      <label className="control-label">Max Color</label>
+                      <ColorSelect
+                        value={opts.heatMaxColor}
+                        onChange={this.updateHeatMaxColor}
+                      />
+                    </div> : null }
+                </div>
+              </div>
+            </Tabs.TabPane> : null }
+          {!includes(['custom', 'heatmap'], opts.globalSeriesType) ?
+            <Tabs.TabPane key="colors" tab="Colors">
+              <ChartColorEditor
+                list={pie ? valuesList : seriesList}
+                options={pie ? opts.valuesOptions : opts.seriesOptions}
+                updateOptions={pie ? this.updateValuesOptions : this.updateSeriesOptions}
+              />
+            </Tabs.TabPane> : null }
+          {opts.globalSeriesType !== 'custom' ?
```

Comment:
These ternary operators probably shouldn't be spread across so many lines, as it makes the code difficult to read and understand. 

I think this is partially a side-effect of JSX allowing full imperative logic intermixed with declarative markup. It somewhat presents a footgun, whereby developers are writing obscure and overly clever code, where a simpler template syntax would suffice.